tests/ui/shared_code_in_if_blocks.rs

```diff
@@ -0,0 +1,198 @@
+fn other_lint() {
+    let x = 10;
+
+    // Only two blocks are the same
+    let _ = if x == 9 {
+        9
+    } else if x == 7 {
+        7
+    } else {
+        7
+    };
+}
+
+fn everything_is_okay() {
+    let x = 10;
+
+    // Single branch
+    if x == 1 {
+        println!("I'm an awesome clippy test")
+    }
+
+    // Same start in only two blocks
+    if x == 10 {
+        println!("x is a value");
+        println!("x is 10");
+    } else if x == 1 {
+        println!("x is 1");
+    } else {
+        println!("x is a value");
+    }
+
+    // Shared code in the middle
+    let _ = if x == 7 {
+        let mut z = 3;
+        println!("x is a value");
+        z *= 2;
+        z
+    } else if x == 100 {
+        let mut y = 1;
+        println!("x is a value");
+        y += 9;
+        y
+    } else {
+        let x = 2;
+        println!("x is a value");
+        x
+    };
+
+    // Returns value with the same name
+    let _ = if x == 9 {
+        let x = 9;
+        println!("--");
+        x
+    } else {
+        let x = 18;
+        println!("^^");
+        x
+    };
+
+    for index in 0..2 {
+        // Okay because we don't have an unconditional else case
+        if index == 6 {
+            println!("Six");
+            continue;
+        } else if index == 9 {
+            println!("Nine");
+            continue;
+        }
+
+        println!("---");
+    }
+}
+
+fn shared_code_at_start() {
+    let x = 17;
+
+    // First statement is the same
+    let _ = if x == 1 {
+        // I'm a comment that shouldn't matter for the lint
+        println!("Hello World");
+
+        4
+    } else {
+        // I'm another comment,
+        // Nice to see you here in the tests
+        println!("Hello World");
+
+        16
+    };
+
+    // First statement is the same
+    if x == 19 {
+        println!("Hello World");
+        let mut y = 9;
+        y += 1;
+        let _z = y;
+    } else {
+        println!("Hello World");
+    };
+
+    if x == 8 {
+        println!("Hello World");
+        println!("How are you today?");
+        println!("From: Test");
+        let _x = 10;
+        let _ = true;
+    } else if x != 2 {
+        println!("Hello World");
+        println!("How are you today?");
+        println!("From: Test");
+        let _x = 10;
+        let _ = false;
+    } else {
+        println!("Hello World");
+        println!("How are you today?");
+        println!("From: Test");
+        let _x = 10;
+    }
+}
+
+fn shared_code_at_end() {
+    let x = 10;
+
+    // Same code at the end
+    if x == 8 {
+        let _ = true;
+        println!("Hello World");
+    } else if x != 2 {
+        let _ = false;
+        println!("Hello World");
+    } else {
+        println!("Hello World");
+    }
+
+    let _z = if x == 8 {
+        println!("Branch 1");
+        let z = 10;
+        foo!();
+        z
+    } else {
+        println!("Branch 2");
+        let z = 10;
+        foo!();
+        z
+    };
```

Comment:
What does the lint do, if you add a slightly different variant of this test:

```rust
    let _z = if x == 8 {
        println!("Branch 1");
        let mut z = 1;
        z += 10;
        foo!();
        z
    } else {
        println!("Branch 2");
        let mut z = 2;
        z += 10;
        foo!();
        z
    };
```

Reply:
The lint would suggest to move the last three lines out of the blocks to prevent code duplication:

<details>
<summary>Lint output</summary>

```
error: All if blocks contain the same code at the end
  --> $DIR/test_file.rs:174:9
   |
LL | /         z += 10;
LL | |         foo!();
LL | |         z
LL | |     } else {
   | |_____^
   |
   = help: Consider moving the code out of the if statement to prevent code duplication
```
</details>

This is an excellent example why this lint is aggressive and why the suggestions are not as automatically applicable. A possible rewrite would be:

```rust
let _z = {
    let mut z = if x == 8 {
        println!("Branch 1");
        1
    } else {
        println!("Branch 2");
        2
    };

    z += 10;
    foo!();
    z
}
```

This is not always the nicest way especially for simple statements like these. This can also be linked to the discussion in #3770 about the `if_same_then_else` what could should actually be moved out. I sadly don't have a better suggestion right now and I believe that this is okay for now since the lint is allowed by default.

Reply:
So this case could be prevented, if we could make sure that:

1. The definition of the assigned variable is before the if-else-block (aka _not in_ the if-else-block) OR 
2. the definition of the assigned variable is contained in the `end` block this lint suggests to move out of the if-else-block.

1\. implies 2. here. So collecting all of the definitions in the if-else and then check if the assigned to variable is in that collection should be a good heuristic.

Reply:
The hard part of implementing this was to decide what should be lined and what not. I see your suggestion and I'll try to implement it :)